File: analysis/model/src/test/java/de/maibornwolff/codecharta/model/PathTest.java
/*
 * Copyright (c) 2017, MaibornWolff GmbH
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *  * Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  * Neither the name of  nor the names of its contributors may be used to
 *    endorse or promote products derived from this software without specific
 *    prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

package de.maibornwolff.codecharta.model;

import org.hamcrest.Matchers;
import org.junit.Test;

import java.util.Arrays;

import static de.maibornwolff.codecharta.model.PathMatcher.matchesPath;
import static junit.framework.TestCase.assertFalse;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.not;
import static org.hamcrest.MatcherAssert.assertThat;

public class PathTest {
    @Test
    public void trivial_path_should_be_equalTo_trivial_path() {
        assertThat(Path.TRIVIAL, matchesPath(Path.TRIVIAL));
    }

    @Test
    public void nonTrivial_path_should_not_be_equalTo_nonTrivial_path() {
        Path nonTrivialPath = new Path(Arrays.asList("bla"));
        assertThat(Path.TRIVIAL, not(matchesPath(nonTrivialPath)));
        assertThat(nonTrivialPath, not(matchesPath(Path.TRIVIAL)));
    }

    @Test
    public void simple_path_should_be_equalTo_simple_path_iff_head_equal() {
        Path nonTrivialPath = new Path(Arrays.asList("bla"));
        Path anotherNonTrivialPath = new Path(Arrays.asList("blubb"));
        assertThat(nonTrivialPath, matchesPath(nonTrivialPath));
        assertThat(nonTrivialPath, not(matchesPath(anotherNonTrivialPath)));
        assertThat(anotherNonTrivialPath, not(matchesPath(nonTrivialPath)));
    }

    @Test
    public void concat_with_trivial_path_should_return_concatinated_path() {
        Path nonTrivialPath = new Path(Arrays.asList("bla"));
        assertThat(Path.TRIVIAL.concat(Path.TRIVIAL), matchesPath(Path.TRIVIAL));
        assertThat(nonTrivialPath.concat(Path.TRIVIAL), matchesPath(nonTrivialPath));
        assertThat(Path.TRIVIAL.concat(nonTrivialPath), matchesPath(nonTrivialPath));
    }

    @Test
    public void complex_path_should_equal_itself() {
        Path firstPath = new Path(Arrays.asList("first"));
        Path expectedPath = new Path(Arrays.asList("second", firstPath.head()));

        assertThat(expectedPath, matchesPath(expectedPath));
    }

    @Test
    public void concat_with_two_simple_paths_should_return_concatinated_path() {
        // given
        Path firstPath = new Path(Arrays.asList("first"));
        Path secondPath = new Path(Arrays.asList("second"));

        // when
        Path concatinatedPath = firstPath.concat(secondPath);

        // then
        Path expectedPath = new Path(Arrays.asList("first", secondPath.head()));
        assertThat(concatinatedPath, not(matchesPath(firstPath)));
        assertThat(concatinatedPath, not(matchesPath(secondPath)));
        assertFalse(concatinatedPath.isSingle());
        assertFalse(concatinatedPath.isTrivial());
        assertThat(concatinatedPath.head(), Matchers.is("first"));
        assertThat(concatinatedPath.tail(), matchesPath(secondPath));
        assertThat(concatinatedPath, matchesPath(expectedPath));
    }


    @Test
    public void last_should_return_last_edged() {
        //given
        String lastEdge = "lastEdge";
        Path path = new Path(Arrays.asList("a", "b", lastEdge));

        // when
        String edge = path.last();

        // then
        assertThat(edge, is(lastEdge));
    }
}

File: analysis/model/src/test/kotlin/de/maibornwolff/codecharta/model/PathTest.kt
package de.maibornwolff.codecharta.model

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test

class PathTest {

    @Nested
    @DisplayName("PathTest > trivial path")
    inner class TrivialPath {
        val trivialPath = Path.TRIVIAL

        @Test
        fun `should behave as expected`() {
            assertThat(trivialPath).isEqualTo(trivialPath)
            assertThat(trivialPath.concat(trivialPath)).isEqualTo(trivialPath)
            assertThat(trivialPath.isTrivial).isTrue
            assertThat(trivialPath.isSingle).isTrue
            assertThat(trivialPath.head).isEqualTo("")
            assertThat(trivialPath.tail.isTrivial).isTrue
        }
    }

    @Nested
    @DisplayName("PathTest > non-trivial path of length one")
    inner class NonTrivialPathLengthOne {
        val edgeName = "bla"
        val nonTrivialPath = Path(edgeName)

        @Test
        fun `should not be equal to trivial path`() {
            assertThat(Path.TRIVIAL).isNotEqualTo(nonTrivialPath)
            assertThat(nonTrivialPath).isNotEqualTo(Path.TRIVIAL)
        }

        @Test
        fun `should not be trivial but be single`() {
            assertThat(nonTrivialPath.isTrivial).isFalse
            assertThat(nonTrivialPath.isSingle).isTrue
        }

        @Test
        fun `should be equal to itself`() {
            assertThat(nonTrivialPath).isEqualTo(nonTrivialPath)
            assertThat(nonTrivialPath).isEqualTo(Path(edgeName))
        }

        @Test
        fun `should not be equal to another non-trivial path`() {
            val anotherNonTrivialPath = Path("blubb")
            assertThat(nonTrivialPath).isNotEqualTo(anotherNonTrivialPath)
        }

        @Test
        fun `concat with trivial path should return same path`() {
            assertThat(nonTrivialPath.concat(Path.TRIVIAL)).isEqualTo(nonTrivialPath)
            assertThat(Path.TRIVIAL.concat(nonTrivialPath)).isEqualTo(nonTrivialPath)
        }

        @Test
        fun `head should be name of single edge`() {
            assertThat(nonTrivialPath.head).isEqualTo(edgeName)
        }

        @Test
        fun `tail should be trivial`() {
            assertThat(nonTrivialPath.tail.isTrivial).isTrue
        }
    }

    @Nested
    @DisplayName("Path Test > non-trivial path of length two")
    inner class NonTrivialPathLengthTwo {
        val firstEdgeName = "first"
        val secondEdgeName = "second"
        val nonTrivialPath = Path(firstEdgeName, secondEdgeName)
        val firstPath = Path(nonTrivialPath.edgesList[0])
        val secondPath = Path(nonTrivialPath.edgesList[1])

        @Test
        fun `should not be equal to trivial path`() {
            assertThat(Path.TRIVIAL).isNotEqualTo(nonTrivialPath)
        }

        @Test
        fun `should not be trivial or single`() {
            assertThat(nonTrivialPath.isTrivial).isFalse
            assertThat(nonTrivialPath.isSingle).isFalse
        }

        @Test
        fun `should equal itself`() {
            assertThat(nonTrivialPath).isEqualTo(nonTrivialPath)
        }

        @Test
        fun `should equal concatenation of its parts`() {
            val concatenatedPath = firstPath.concat(secondPath)

            assertThat(nonTrivialPath).isEqualTo(concatenatedPath)
        }

        @Test
        fun `should not be equal to its parts`() {
            assertThat(nonTrivialPath).isNotEqualTo(firstPath)
            assertThat(nonTrivialPath).isNotEqualTo(secondPath)
        }

        @Test
        fun `head should be first edge name`() {
            assertThat(nonTrivialPath.head).isEqualTo(firstEdgeName)
        }

        @Test
        fun `tail should be second edge`() {
            assertThat(nonTrivialPath.tail).isEqualTo(secondPath)
        }
    }

    @Test
    fun `fitting edges from tail with should calculate correctly`() {
        val paths = listOf(
            Path(),
            Path("1"),
            Path("1", "2"),
            Path("1", "2", "3"),
            Path("a"),
            Path("1", "a"),
            Path("0", "1", "a"),
            Path("1", "2", "a"),
            Path("0", "1", "2", "a")
        )

        assertThat(paths.map { paths[0].fittingEdgesFromTailWith(it) })
            .isEqualTo(listOf(0, 0, 0, 0, 0, 0, 0, 0, 0))

        assertThat(paths.map { paths[4].fittingEdgesFromTailWith(it) })
            .isEqualTo(listOf(0, 0, 0, 0, 1, 1, 1, 1, 1))

        assertThat(paths.map { paths[5].fittingEdgesFromTailWith(it) })
            .isEqualTo(listOf(0, 0, 0, 0, 1, 2, 2, 1, 1))

        assertThat(paths.map { paths[7].fittingEdgesFromTailWith(it) })
            .isEqualTo(listOf(0, 0, 0, 0, 1, 1, 1, 3, 3))
    }
}
